### PR TITLE
SI-7753 InstantiateDependentMap narrows type of unstable args

### PR DESCRIPTION
**Pull request by adriaanm**
[Most of this comment and the initial fix were implemented by Jason Zaugg.
I just cleaned it up a bit.]

After a soundness fix in SI-3873, instantiation of dependent
method type results behaved differently depending on whether the argument
from which we were propagating information had a stable type
or not. This is particular to substitution into singleton types
over the parameter in question.

If the argument was stable, it was substituted into singleton
types, such as the one below in the prefix in `a.type#B`
(which is the longhand version of `a.B`)

  scala> class A { type B >: Null <: AnyRef }
  defined class A

  scala> object AA extends A { type B = String }
  defined object AA

  scala> def foo(a: A): a.B = null
  foo: (a: A)a.B

  scala> foo(AA)
  res0: AA.B = null

But what if it isn't stable?

  scala> foo({def a = AA; a: A { type B <: String}})
  res1: a.B = null

This commit changes that to:

  scala> foo({def a = AA; a: A { type B <: String}})
  res1: A{type B <: String}#B = null

review by @retronym

### CODE REVIEW

**Comment by adriaanm** [comment]
This continues where #3482 left off.

**Comment by paulp** [comment]
No, I don't mean "fix it", I mean put it in as a neg test which should eventually be pos.

File: test/files/pos/t7753.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    val converted = conv(foo)
+
+    val v1: Int = indirect(converted)(23)  // Okay (after refining the return type `instance` in the return type of `conv`)
+    /*
+    indirect(converted){(v: converted.instance.Out)converted.instance.Out}(
+      23{Int(23)}
+    ){converted.instance.Out};
+    */
+
+    val v2: Int = indirect(conv(foo))(23)  // Fails.
```
Comment (clhodapp) [inline test/files/pos/t7753.scala]: You might want to rebase the comment out as it's a bit misleading now that this is a pos test.

**Comment by paulp** [comment]
I'll send you a PR. That test is freaking poetry.

**Comment by paulp** [comment]
#3508

**Comment by adriaanm** [comment]
test comment improved

**Comment by retronym** [comment]
LGTM.

**Comment by adriaanm** [comment]
Will rebase on master once #3516 is merged

**Pull request closed without merging**